app/serializers/api/address_serializer.rb

```diff
@@ -4,7 +4,11 @@ class Api::AddressSerializer < ActiveModel::Serializer
 
   attributes :id, :zipcode, :city, :state_name, :state_id,
              :phone, :firstname, :lastname, :address1, :address2, :city, :country_id,
-             :zipcode
+             :zipcode, :country_name
+
+  def country_name
+    object.country.andand.name
```

Comment:
I don't think we are eager loading the country so this will issue an extra query to get it while a single one with a join would do. Please, check that out.

This with `has_many`  declarations in serializers is one of the most common causes of N+1s we have in the app.

Reply:
I created this section in the code conventions:
https://github.com/openfoodfoundation/openfoodnetwork/wiki/Code-Conventions#performance

Can you help me write something on the "how to avoid this"?

do you know what's the code I should right in each case?

Reply:
I found [this option](https://stackoverflow.com/a/18134650/1812272). But this is only for nested attributes like object.line_items.variants. Not sure how to make the object itself eagerload for example adjustments...
but there must be a way and definitely something we should do in serializers.

Reply:
we should instead call `#includes` from the controller (or whatever abstraction might be dealing with the data loading in each case) and pass on the data to the view layer. When leave that to perform the data loading is when N+1s pop-up.

In this cases the following patch

```rb
diff --git a/app/controllers/api/orders_controller.rb b/app/controllers/api/orders_controller.rb
index 03d4e8caa..bbbde5fe0 100644
--- a/app/controllers/api/orders_controller.rb
+++ b/app/controllers/api/orders_controller.rb
@@ -26,7 +26,10 @@ module Api
     end
 
     def order
-      @order ||= Spree::Order.find_by_number!(params[:id])
+      @order ||= Spree::Order
+        .where(number: params[:id])
+        .includes(line_items: :variant)
+        .first
     end
   end
 end
```

makes

```
openfoodnetwork@ofn:/opt/openfoodnetwork$ tail -n 200 log/development.log | grep spree_variants                                                                                              
  Spree::Variant Load (0.5ms)  SELECT "spree_variants".* FROM "spree_variants" WHERE "spree_variants"."id" = 6 LIMIT 1                                                                       
  Spree::Variant Load (0.5ms)  SELECT "spree_variants".* FROM "spree_variants" WHERE "spree_variants"."id" = 4 LIMIT 1                                                                       
  Spree::Variant Load (0.4ms)  SELECT "spree_variants".* FROM "spree_variants" WHERE "spree_variants"."id" = 8 LIMIT 1                                                                       
  Spree::Variant Load (0.4ms)  SELECT "spree_variants".* FROM "spree_variants" WHERE "spree_variants"."id" = 2 LIMIT 1     
```

become 

```
openfoodnetwork@ofn:/opt/openfoodnetwork$ tail -n 200 log/development.log | grep spree_variants
  Spree::Variant Load (0.5ms)  SELECT "spree_variants".* FROM "spree_variants" WHERE "spree_variants"."id" IN (6, 4, 8, 2) AND ("spree_variants".deleted_at IS NULL)  
```

Note however that as this is a `show` action we do perform queries like

```
SELECT "spree_line_items".* FROM "spree_line_items" WHERE "spree_line_items"."order_id" IN (4) ORDER BY spree_line_items.created_at ASC
```

for some of the involved tables but perform N+1 for tables related to variants and products, as well as those. I see things like 

```
SELECT "spree_stock_items".* FROM "spree_stock_items" WHERE "spree_stock_items"."variant_id" = 2 ORDER BY id ASC LIMIT 1
```

That should also be eager loaded by using `#includes`.

Reply:
ah, this is amazing!

This removes N+1 queries on line_items, variants, products, stock_items and prices :-O
```ruby
   @order ||= Spree::Order.
        where(number: params[:id]).
        includes(line_items: { variant: [:product, :stock_items, :default_price] }).
        first
```

I tried stock_locations but it didnt work, I cant see where the active scope is triggered for stock_locations.

Reply:
I have not managed to act on the initial point about countries...
I tried includes with
{ ship_address: :country }
but it doesnt work...